File: ueeventdumper.py
#!/usr/bin/env python3
import sys
import os
import json

def extract_sound_name(entry):
    if not entry:
        return "Unknown"

    props = entry.get("Properties", {})

    for key in ("Event_FP", "Event_TP"):
        if key in props and isinstance(props[key], dict):
            n = props[key].get("ObjectName")
            if n and n != "None":
                return n

    return entry.get("Name", "Unknown")


def process_json(path):
    with open(path, "r", encoding="utf-8") as f:
        data = json.load(f)

    anim = next((i for i in data if i.get("Type") in ("AnimSequence","AnimMontage")), None)
    if not anim:
        return False, "AnimSequence/AnimMontage not found"

    props = anim.get("Properties", {})
    notifies = props.get("Notifies", [])
    num_frames = props.get("NumFrames", 0)
    seq_length = props.get("SequenceLength", 1)

    fps = (num_frames / seq_length) if seq_length != 0 else 0

    sound_map = {
        f"AnimNotify_WeaponSound'{e.get('Outer')}:{e.get('Name')}'": e
        for e in data if e.get("Type") == "AnimNotify_WeaponSound"
    }

    result = []
    for n in notifies:
        event_type = n.get("NotifyName", "Unknown")
        t = n.get("Time", n.get("LinkValue", 0))
        frame = int(t * fps)

        if event_type == "WeaponSound":
            ref = n.get("Notify", {}).get("ObjectName", "")
            name = extract_sound_name(sound_map.get(ref))
        else:
            name = event_type

        result.append(f"{t:.6f}\t{frame}\t{event_type}\t{name}")

    if not result:
        return False, "No notifies found"

    out = os.path.splitext(path)[0] + "_events.txt"
    with open(out, "w", encoding="utf-8") as f:
        f.write("Time(sec)\tFrame\tEvent\tDetails\n")
        f.write("\n".join(result))

    return True, f"Saved → {out}"


def main():
    if len(sys.argv) < 2:
        print("\nDrag & Drop JSON onto this EXE\n")
        return

    path = sys.argv[1]
    if not os.path.exists(path) or not path.lower().endswith(".json"):
        print("Invalid or non-json file.")
        return

    ok, msg = process_json(path)
    print(msg)


if __name__ == "__main__":
    main()

File: ueeventdumper.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <unordered_map>
#include <iomanip>
#include "json.hpp"

using json = nlohmann::json;

std::string extract_sound_name(const json* entry) {
    if (!entry || entry->is_null()) return "Unknown";

    if (entry->contains("Properties") && entry->at("Properties").is_object()) {
        const json& props = entry->at("Properties");
        for (const std::string& key : {"Event_FP", "Event_TP"}) {
            if (props.contains(key) && props[key].is_object()) {
                std::string n = props[key].value("ObjectName", "");
                if (!n.empty() && n != "None") return n;
            }
        }
    }
    return entry->value("Name", "Unknown");
}

std::pair<bool, std::string> process_json(const std::string& path) {
    std::ifstream f(path);
    if (!f.is_open()) return {false, "Cannot open file"};

    json data;
    try {
        f >> data;
    } catch (...) {
        return {false, "Failed to parse JSON"};
    }
    f.close();

    const json* anim = nullptr;
    for (const auto& item : data) {
        std::string type = item.value("Type", "");
        if (type == "AnimSequence" || type == "AnimMontage") {
            anim = &item;
            break;
        }
    }
    if (!anim) return {false, "AnimSequence/AnimMontage not found"};

    const json& props = (*anim)["Properties"];
    const json& notifies = props.value("Notifies", json::array());
    double num_frames = props.value("NumFrames", 0.0);
    double seq_length = props.value("SequenceLength", 1.0);
    double fps = (seq_length != 0) ? num_frames / seq_length : 0.0;

    std::unordered_map<std::string, const json*> sound_map;
    for (const auto& e : data) {
        if (e.value("Type", "") == "AnimNotify_WeaponSound") {
            std::string key = "AnimNotify_WeaponSound'" + e.value("Outer", "") + ":" + e.value("Name", "") + "'";
            sound_map[key] = &e;
        }
    }

    std::vector<std::string> result;
    for (const auto& n : notifies) {
        std::string event_type = n.value("NotifyName", "Unknown");
        double t = n.value("Time", n.value("LinkValue", 0.0));
        int frame = static_cast<int>(t * fps);

        std::string name;
        if (event_type == "WeaponSound") {
            std::string ref = n.value("Notify", json::object()).value("ObjectName", "");
            name = extract_sound_name(sound_map.count(ref) ? sound_map[ref] : nullptr);
        } else {
            name = event_type;
        }

        std::ostringstream line;
        line << std::fixed << std::setprecision(6)
             << t << "\t" << frame << "\t" << event_type << "\t" << name;
        result.push_back(line.str());
    }

    if (result.empty()) return {false, "No notifies found"};

    std::string out = path.substr(0, path.find_last_of('.')) + "_events.txt";
    std::ofstream fo(out);
    for (const auto& line : result) fo << line << "\n"; // без заголовка
    fo.close();

    return {true, "Saved → " + out};
}

int main(int argc, char* argv[]) {
    if (argc < 2) {
        std::cout << "\nDrag & Drop JSON onto this EXE\n";
        return 0;
    }

    std::string path = argv[1];
    std::ifstream check(path);
    if (!check.good() || path.size() < 5 || path.substr(path.size() - 5) != ".json") {
        std::cout << "Invalid or non-json file.\n";
        return 1;
    }
    check.close();

    auto [ok, msg] = process_json(path);
    std::cout << msg << "\n";

    return 0;
}